staging/src/k8s.io/client-go/tools/watch/retrywatcher.go

```diff
@@ -0,0 +1,183 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package watch
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/golang/glog"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+// WatcherFunc is a function that is responsible for creating a watcher starting at sinceResourceVersion.
+type WatcherFunc func(sinceResourceVersion string) (watch.Interface, error)
+
+// resourceVersionGetter is an interface used to get resource version from events.
+// We can't reuse an interface from meta otherwise it would be a cyclic dependency and we need just this one method
+type resourceVersionGetter interface {
+	GetResourceVersion() string
+}
+
+// RetryWatcher will make sure that in case the underlying watcher is closed (e.g. due to API timeout or etcd timeout)
+// it will get restarted from the last point without the consumer even knowing about it.
+// RetryWatcher does that by inspecting events and keeping track of resourceVersion.
+// Especially useful when using watch.Until where premature termination is causing troubles and flakes.
+// Please note that this is not resilient to ETCD cache not having the resource version anymore - you would need to
+// use Informers for that.
+type RetryWatcher struct {
+	lastResourceVersion string
+	watcherFunc         WatcherFunc
+	resultChan          chan watch.Event
+	stopChan            chan struct{}
+	doneChan            chan struct{}
+}
+
+// NewRetryWatcher creates a new RetryWatcher.
+// It will make sure that watcher gets restarted in case of recoverable errors.
+// The initialResourceVersion will be given to watchFunc when first called.
+func NewRetryWatcher(initialResourceVersion string, watcherFunc WatcherFunc) *RetryWatcher {
+	rw := &RetryWatcher{
+		lastResourceVersion: initialResourceVersion,
+		watcherFunc:         watcherFunc,
+		stopChan:            make(chan struct{}),
+		doneChan:            make(chan struct{}),
+		resultChan:          make(chan watch.Event, 0),
+	}
+
+	go rw.receive()
+	return rw
+}
+
+func (rw *RetryWatcher) send(event watch.Event) bool {
+	// Writing to an unbuffered channel is blocking and we need to check if we need to be able to stop while doing so!
+	select {
+	case rw.resultChan <- event:
+		return true
+	case <-rw.stopChan:
+		return false
+	}
+}
+
+// doReceive returns true when it is done, false otherwise
+func (rw *RetryWatcher) doReceive() bool {
+	watcher, err := rw.watcherFunc(rw.lastResourceVersion)
+	if err != nil {
+		_ = rw.send(watch.Event{
+			Type:   watch.Error,
+			Object: apierrors.NewInternalError(fmt.Errorf("RetryWatcher: watcherFunc failed: %v", err)).Status(),
```

Comment:
I dunno if this is an artifact of my local setup or not (I'm using k8s.io/apimachinery from kubernetes-1.12.2 and Go 1.10.3), but this is giving me:

```
...: cannot use ".../vendor/k8s.io/apimachinery/pkg/api/errors".NewInternalError(fmt.Errorf("RetryWatcher: watcherFunc failed: %v", err)).Status() (type ".../vendor/k8s.io/apimachinery/pkg/apis/meta/v1".Status) as type runtime.Object in field value:
	".../vendor/k8s.io/apimachinery/pkg/apis/meta/v1".Status does not implement runtime.Object (DeepCopyObject method has pointer receiver)
```

Instead, I had to use:

```go
status := apierrors.NewInternalError(fmt.Errorf("RetryWatcher: watcherFunc failed: %v", err)).Status()
_ = rw.send(watch.Event{
  Type:   watch.Error,
  Object: &status,
})
```

Reply:
Also, can we have `fmt.Errorf("retry watcher:..."` to keep golint from complaining about:

    error strings should not be capitalized or end with punctuation or a newline

for low `-min_confidence`?  We just need something to give some context; we don't need to quote the Go type directly.

Reply:
old apimachinery; I can make the error start with lowercase